File: codeql-custom-queries-java/android/prefer-lazy-loading-compose.java
class ListView {
    void setAdapter(Object adapter) {}
    void scrollTo(int x, int y) {}
}

class GridView {
    void setAdapter(Object adapter) {}
    void setNumColumns(int numColumns) {}
}

class RecyclerView {
    void setAdapter(Object adapter) {}
    void setLayoutManager(Object layoutManager) {}
    void scrollToPosition(int position) {}
}

class NoncompliantLazyLoading {

    // 🚫 Noncompliant - ListView field
    ListView listView = null; // $ Alert

    // 🚫 Noncompliant - GridView field
    GridView gridView = null; // $ Alert

    // 🚫 Noncompliant - RecyclerView field
    RecyclerView recyclerView = null; // $ Alert

    void setupList() {
        // 🚫 Noncompliant - local ListView
        ListView list = new ListView(); // $ Alert
        list.setAdapter(null);
    }

    void setupGrid() {
        // 🚫 Noncompliant - local GridView
        GridView grid = new GridView(); // $ Alert
        grid.setNumColumns(3);
    }

    void setupRecycler() {
        // 🚫 Noncompliant - local RecyclerView
        RecyclerView recycler = new RecyclerView(); // $ Alert
        recycler.setLayoutManager(null);
        recycler.setAdapter(null);
    }

    void setupMultiple() {
        // 🚫 Noncompliant - multiple views in same method
        ListView l = null;      // $ Alert
        GridView g = null;      // $ Alert
        RecyclerView r = null;  // $ Alert
    }
}

class CompliantLazyLoading {
    // ✅ Compliant - no classic Android views
    void doNothing() {}
}

File: codeql-custom-queries-java/android/prefer-lazy-loading-compose.kt
// Stubs
class ListView
class GridView
class RecyclerView
class LazyColumn
class LazyRow
class LazyVerticalGrid
class LazyHorizontalGrid

// 🚫 Noncompliant - using eager loading views
class NoncompliantLazyLoading {

    val listView: ListView? = null         // $ Alert
    val gridView: GridView? = null         // $ Alert
    val recyclerView: RecyclerView? = null // $ Alert

    fun setupViews() {
        val list: ListView? = null         // $ Alert
        val grid: GridView? = null         // $ Alert
        val recycler: RecyclerView? = null // $ Alert
    }
}

// ✅ Compliant - using Jetpack Compose lazy components
class CompliantLazyLoading {

    val lazyColumn: LazyColumn? = null         // OK
    val lazyRow: LazyRow? = null               // OK
    val lazyGrid: LazyVerticalGrid? = null     // OK
    val lazyHGrid: LazyHorizontalGrid? = null  // OK
}